ReadMe.txt
/*
 * README
 *
 * T38FAX Pseudo Modem
 *
 * Original author: Vyacheslav Frolov
 *
 * Additional contributors: Henning Holtschneider, Pete Davidson
 *
 */

1. Introduction
---------------

What is t38modem?

From your fax or voice application view point it's a fax/voice modem pool.
From IP network view point it's a H.323/SIP endpoint with T.38 fax support.
From your view point it's a gateway between an application and IP network.

The homepage for t38modem project is http://t38modem.sourceforge.net/.

T38modem version 2.0.0 can only be used with opal/ptlib revision 24174.  

To use later versions of opal/ptlib see the t38modem projects at github:
https://github.com/hehol/t38modem
https://github.com/PeteDavidson/t38modem

2. Building
-----------

2.1. Building for Unix
----------------------

Building with Open Phone Abstraction Library (OPAL):

  $ export PTLIBDIR=$path_to_libs/ptlib
  $ export OPALDIR=$path_to_libs/opal
  $ make USE_OPAL=1 opt

2.2. Building for Windows
-------------------------

Building with Open Phone Abstraction Library (OPAL):

  Start Microsoft Visual C++ 2005 with opal\t38modem_2005.vcproj file.
  Set Active Configuration to "t38modem - Win32 Release".
  Build t38modem.exe.

3. Examples
-----------

To get info about t38modem command line syntax enter:

  $ t38modem --help


3.1. Starting
-------------

Starting with Open H323 Library or H323 Plus Library:

  $ t38modem -n -o trace.log -p ttyx0,ttyx1 --old-asn \
                               --route 0@127.0.0.1 --route all@172.16.33.21

This will create two modems (/dev/ttyx0 and /dev/ttyx1) and H.323 endpoint.
If dialed number begins with '0' then it will be routed to a local host
(leading '0' will be discarded). Other dialed numbers will be routed to
172.16.33.21.

Starting with OPAL:

  $ t38modem -n -o trace.log -p ttyx0,ttyx1 \
                               --route "modem:0.*=h323:<dn!1>@127.0.0.1" \
                               --route "modem:1.*=sip:<dn>@172.16.33.20" \
                               --route "modem:2.*=h323:<dn>@172.16.33.21" \
                               --route "h323:.*=modem:<dn>" \
                               --route "sip:.*=modem:<dn>"

This will create two modems (/dev/ttyx0 and /dev/ttyx1) and H.323 and SIP endpoints.
If dialed number begins with '0' then it will be routed to a local host.
If dialed number begins with '1' then it will be routed to SIP endpoint 172.16.33.20.
If dialed number begins with '2' then it will be routed to H.323 endpoint 172.16.33.21.
Leading '0', '1' and '2' will be discarded.

Q. I try to use T38modem, but after run "t38modem -p ttyx0" I get a message
   "Could not open /dev/ptyx0: No such file or directory".
A. Looks like you don't have legacy PTY devices compiled in your kernel.
   You need to re-compile the kernel with 'Legacy PTY Support'.
   Alternatively, you can build t38modem with USE_UNIX98_PTY=1 option and use
   -p +/dev/ttyx0,+/dev/ttyx1 instead of -p ttyx0,ttyx1.

FreeBSD Users: You need to use  -p ttypa,ttypb instead of -p ttyx0,ttyx1.
               Remember to replace ttyx0 with ttypa and ttyx1 with ttypb
               when following the rest of these instructions.
               This will create two modems /dev/ttypa and /dev/ttypb.

Windows Users: You need two COM ports connected via Null-modem cable to create one modem.
               If your COM1 connected to COM2 and COM3 connected to COM4 you need to use
               -p \\.\COM2,\\.\COM4 instead of  -p ttyx0,ttyx1.
               This will create two modems COM1 and COM3.
               Q. How to use t38modem without additional COM port hardware on Windows?
               A. Replace a pair of physical COM ports with a pair of virtual COM ports.
                  See http://com0com.sourceforge.net/ project for details.
               Q. What model of modem to select in Add Hardware Wizard?
               A. Select "Standard 1440 bps Modem".

Cisco Users:   Possible additionaly you will need to use --h245tunneldisable option.

3.2. Testing (you need two consoles)
------------------------------------
(FreeBSD users - remeber to use /dev/ttypa and /dev/ttypb with 'cu -l')
(Windows users - use COM1 and COM3 with HyperTerminal)

$ cu -l /dev/ttyx0	    $ cu -l /dev/ttyx1
Connected.		    Connected.
<-- at		            <-- at
--> OK                      --> OK
                            (wait at least 10 secs)
<-- atdt012345

                            -->
                            --> RING
                            -->
                            --> RING
                            <-- ati9
                            --> NDID = 12345
                            --> OK
                            -->
                            --> RING
                            -->
                            --> RING
                            <-- ata
--> CONNECT                 --> CONNECT
<-- x
--> OK
<-- ath
--> OK
                            -->
                            --> ERROR
<-- at
--> OK
                            <-- at
                            --> OK
...                         ...


3.3. Example of Cisco config (loopback)
---------------------------------------

10.0.2.12 --> Cisco port 2:D --E1-cable--> Cisco port 3:D --> 10.0.2.12

dial-peer voice 3340 voip
 incoming called-number 3334....
 codec g711alaw
 fax rate 14400
 fax protocol t38 ls-redundancy 0 hs-redundancy 0

dial-peer voice 3341 pots
 destination-pattern 3334....
 port 2:D
 forward-digits 7

dial-peer voice 3342 pots
 incoming called-number 334....
 direct-inward-dial
 port 3:D
exit

dial-peer voice 3343 voip
 destination-pattern 334....
 session target ipv4:10.0.2.12
 codec g711alaw
 fax rate 14400
 fax protocol t38 ls-redundancy 0 hs-redundancy 0
exit

3.4. Example of HylaFAX modem config files
------------------------------------------

Copy HylaFAX/etc/config.ttyx to HylaFAX's etc directory

Create simbolic links:

config.ttyx0 -> config.ttyx
config.ttyx1 -> config.ttyx

Start HylaFAX with new modems:

$ .../faxgetty -D ttyx0
$ .../faxgetty -D ttyx1

(FreeBSD users - don't forget we are using ttypa and ttypb)

4. AT commands specific to t38modem
-----------------------------------

4.1. AT#CID command
-------------------

4.1.1 Reporting Caller ID and DID after the first RING
------------------------------------------------------

#CID=0	- disable Caller ID and DID reporting (default).
#CID=1	- enable only Caller ID reporting.
#CID=10	- enable Caller ID and DID reporting.
#CID=11	- enable only DID reporting.

Example:

<-- AT#CID=10
--> OK
--> RING
--> DATE = <MMDD>              ; Caller ID
--> TIME = <HHMM>              ; Caller ID
--> NMBR = <calling number>    ; Caller ID
--> NAME = <calling name>      ; Caller ID
--> NDID = <called number>     ; DID
--> RING
--> RING

4.2. ATI command
----------------

4.2.1 calling/called number reporting
-------------------------------------

I8	- reports calling number for last incoming call.
I9	- reports called number for last incoming call.

Example:

<-- ATI8I9
--> NMBR = <calling number>
--> NDID = <called number>
--> OK

4.3. ATD command
----------------

4.3.1 Fax mode modifiers
------------------------

F - force fax mode (T.38 or G.711 pass-trough) after dialing
    (with OPAL can be overridden by route option OPAL-Force-Fax-Mode=false).
V - do not force fax mode after dialing (default)
    (with OPAL can be overridden by route option OPAL-Force-Fax-Mode=true).

Examples:

<-- ATDF<user's number>

    force fax mode after dialing but user can override it by
    inserting V into <user's number>.

<-- ATD<user's number>V

    do not force fax mode after dialing and user can't override it
    by inserting F into <user's number>.

4.3.2 calling/called number modifiers
-------------------------------------

L     - purge calling number and start dialing it.
T,P,D - continue to dial of called number if dialing of calling number was started.

If calling number is empty after processing ATD command then t38modem's
local party number used.

The calling number is a string of 0 or more of the characters:

  "0 1 2 3 4 5 6 7 8 9 * #"

Note: dialing of calling number is not allowed in online command state.

The called number is a string of 0 or more of the characters:

  "0 1 2 3 4 5 6 7 8 9 * # + A B C D"

Note: '+' will be ignored in online command state or after '@' modifier.

Examples:

<-- ATD<user's number>

    calling number is t38modem's local party number but user can override
    it by inserting L<user's calling number> into <user's number>

<-- ATD<user's number>L

    calling number is t38modem's local party number and user can't override
    it by inserting L<user's calling number> into <user's number>

<-- ATDL<calling number>T<user's number>

    calling number is <calling number> but user can override
    it by inserting L<user's calling number> into <user's number>

<-- ATD<user's number>L<calling number>

    calling number is <calling number> and user can't override
    it by inserting L<user's calling number> into <user's number>

4.3.3 Wait for answer modifier
------------------------------

@ - dial the called number part collected before first @, wait till the call
    answered and play the called number part collected after last @.
    Implicitly '@' will continue dialing of called number (see 4.3.2).

Examples:

<-- ATS8?
--> 002
--> OK
<-- ATD12345@,,,4321

    dial number 12345, wait till the call answered, wait 6 seconds
    (3 commas * 2 seconds), play digits 4321.

<-- ATD12345@,,,4321@

    dial number 12345.

4.4. AT#DFRMC command
---------------------

4.4.1 Set delay for CONNECT result code for AT+FRM command
----------------------------------------------------------

#DFRMC=0	- disable delay (default).
#DFRMC=25	- set delay to 250 ms.

4.5. AT#HCLR command
--------------------

4.5.1 Set clear call mode for ATH command
-----------------------------------------

#HCLR=0	- ATH command will not clear not answered incoming call (default).
#HCLR=1	- ATH command will clear not answered incoming call.

4.6. AT#CIDFMT command
----------------------

AT#CIDFMT command allows to set value format for NAME and NMBR tags of Caller ID.

Syntax:

  #CIDFMT=[<name format>][,[<nmbr format>]]

If a format element is empty then its value will be untouched.

Subparameter <name format>:

  0   - report "NAME = <calling name>" (default)
  1   - report "NAME = "
  2   - report "NAME = <called number>"
  3   - report "NAME = <called number> <- <calling name>"

Subparameter <nmbr format>:

  0   - report "NMBR = <calling number>" (default)
  1   - report "NMBR = "
  2   - report "NMBR = <called number>"
  3   - report "NMBR = <called number>#<calling number>"



Changes.txt
/*
 * CHANGES
 *
 * T38FAX Pseudo Modem
 *
 * Original author: Vyacheslav Frolov
 *
 */

Changelog for t38modem 3.14.0 (Mar 3 2016)
--- Requires Opal 3.10.14 / Ptlib 2.10.14
* Additional debug on SIP Registration.
* New option: --sip-retry-403-forbidden
  UNIFY systems will respond 403 Forbidden when resetting to apply new
  settings. RFC 3261 says this is a final answer and the request should
  not be retried. UNIFY says that is the way they work and you have to 
  retry. This violates ithe RFC, but is needed to make UNIFY work.
* Added ability to use the all of the Fake Codecs:
    G.722, G.722.1, G.722.2, G.726-40K, G.726-32K, G.726-24K, G.726-16K,
    G.728, G.729, G.729A, G.729B, G.729A/B, G.723.1, G.723.1(5.3k),
    G.723.1A(6.3k), G.723.1A(5.3k), GSM-06.10, GSM-AMR, iLBC
  Specify with --fake-audio option and --sip-audio or --h323-audio options.  

Changelog for t38modem 3.13.0 (Aug 10 2015)
--- Requires Opal 3.10.13 / Ptlib 2.10.13
* Opal 3.10.13 changed to fix Cisco CUBE bug where it starts with sequenc
  number 32768 instead of the required 0 in UDPTL T.38 packets.
* Added missing DTMF method in SDP on Invite. Fixes some Alcatel systems.

Changelog for t38modem 3.12.1 (Jan 31 2015)
--- Requires Opal 3.10.12 / Ptlib 2.10.12
* New option: --force-fax-mode-delay

Changelog for t38modem 3.12.0 (Jan 31 2015)
--- Requires Opal 3.10.12 / Ptlib 2.10.12
* Supports G.711 Passthru mode with the addition of spandsp 0.0.6 plugin.
* Changed port numbers to work on Verszon network. We now use 10000-10999.

Changelog for t38modem 3.11.0 (Nov 12 2013)
* Changes to work with Opal 3.10.11.

Changelog for t38modem 2.0.1 (Sep 24 2013)
* Busy Fix (When busy set cause code to EndedByLocalBusy instead of 
  EndedByLocalUser)

Changelog for t38modem 2.0.0 (Apr 13 2011)

* (OPAL) Implemented G.711 fallback to fax pass-through mode.
* (OPAL) Added ability to disable T.38 fax mode.
* (OPAL) Implemented multiple per-route options to override command line
  options (NOTE: the route comma delimiter in command line was deprecated
  for this reason).
* Added ability to set Q.931 bearer capability information element for
  H.323 connections.
* Enabled dialing characters # and * in extension number.
* Replaced hardcoded timeouts by S7 value (V.250).
* (OPAL) Implemented fake audio codecs G.722, G.722.1, G.722.2, 
  G.723.1, G.723.1(5.3k), G.723.1A(5.3k), G.723.1A(6.3k), G.726-16K,
  G.726-24K, G.726-32K, G.726-40K, G.728, G.729, G.729A, G.729A/B,
  G.729B, GSM-06.10, GSM-AMR, iLBC.
* (OPAL) Restored and improved ability to set error recovery redundancy for
  IFP packets.
* (OPAL) Restored and improved ability to continuously re-send last UDPTL
  packet.
* (OPAL) Implemented fallback to alternate route option.
* (OPAL) Added ability to build with OPAL configured to disable H.323 or
  SIP protocol support.
* Allowed "+ A B C D" dialing digits in called number.
* Added ignoring unrecognized dialing digits (V.250).
* Implemented dial string terminated by a semicolon ("ATD<dialstring>;[...]").
* Implemented fake Ring Tone and Busy Tone generation for voice-modem mode.
* Replaced hardcoded workaround for mgetty-voice by conditional-compiling one
  (ALAW_132_BIT_REVERSE).
* Added ability to put called number to Caller ID info with NAME or NMBR tags.
* Fixed some other bugs and incompatibilities. Made some other improvements.
  Added some other options.

Changelog for t38modem 1.2.0 (Oct  8 2009)

* Enabled dialing characters # and *.
* Added ability to dial extension numbers.
* Dial modifiers 'T' and 'P' can be used instead 'D'.
* Fixed incompatibilities with some voice applications (V.253).
* Added ability to dial after call answer.
* Added semicolon concatenating of AT commands.
* Added ability to dial and switch to answering mode ("ATD<dialstring>;A").
* Fixed incompatibilities with some OS.
* (OPAL) Added ability to use route pattern "modem:.*@tty".
* (OPAL) Added ability to enable/disadle/reorder audio formats.
* (OPAL) Removed --h323-old-asn and --sip-old-asn options.
* Fixed some other bugs.

Changelog for t38modem 1.1.0 (Apr 16 2009)

* Fixed 100% CPU usage at the begining of T.38 session.
* Fixed transmission fail with ECM.
* Added H323Plus support.
* Added OPAL v3.6 Lalande support.
* (OPAL) Added STUN support.
* (OPAL) Added ability to set outbound proxy for SIP.
* (OPAL) Added ability to set registrar for SIP.
* (OPAL) options --h323-redundancy, --h323-repeat, --sip-redundancy
  and --sip-repeat are disabled (because no support in OPAL).

Oct  6 2008 - t38modem moved to http://sourceforge.net/projects/t38modem/.

Changelog for t38modem 1.0.0 (Aug 29 2008)

* Added OPAL support.
* Implemented voice mode functionality.
* The code optimized for original ASN.1 sequence in T.38 (06/98)
  Annex A (it's optimal if you use --old-asn option). Compiling
  with OPTIMIZE_CORRIGENDUM_IFP will optimize for CORRIGENDUM No. 1
  fix (it's optimal if you do not use --old-asn option).
* Added ability to continuously re-send last UDPTL packet.
* Added Unix98 PTY support.

Changelog for t38modem 0.8.4 (Mar 22 2007)

* Fixed incompatibilities with some endpoints.
* Fixed big file descriptors problem.
* Fixed some possible dead locks.
* Implemented AT#HCLR command.
* Added ability to set range of IP ports to use.

  NOTE: To compile t38modem since version 0.8.4 you need
        openh323 since version 1.19.1.

Changelog for t38modem 0.8.3 (Jun 29 2006)

* Fixed race condition on start up.
* Added -v option to keep Debian packaging happy.

Changelog for t38modem 0.8.2 (Sep 14 2005)

* Fixed incompatibilities with some endpoints.
* Fixed UDPTL size.

Changelog for t38modem 0.8.1 (Mar 31 2005)

* Fixed incompatibilities with some fax applications.
* Implemented some AT commands.
* Fixed some bugs.

Changelog for t38modem 0.8.0 (Oct 29 2004)

* Ported to Windows.

Changelog for t38modem 0.7.0 (Jul  7 2004)

* Implemented FCS generation. The DCE transfers FCS octets after
  the HDLC data to the DTE (as it should be).

g711.c
/*
 * This source code is a product of Sun Microsystems, Inc. and is provided
 * for unrestricted use.  Users may copy or modify this source code without
 * charge.
 *
 * SUN SOURCE CODE IS PROVIDED AS IS WITH NO WARRANTIES OF ANY KIND INCLUDING
 * THE WARRANTIES OF DESIGN, MERCHANTIBILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE, OR ARISING FROM A COURSE OF DEALING, USAGE OR TRADE PRACTICE.
 *
 * Sun source code is provided with no support and without any obligation on
 * the part of Sun Microsystems, Inc. to assist in its use, correction,
 * modification or enhancement.
 *
 * SUN MICROSYSTEMS, INC. SHALL HAVE NO LIABILITY WITH RESPECT TO THE
 * INFRINGEMENT OF COPYRIGHTS, TRADE SECRETS OR ANY PATENTS BY THIS SOFTWARE
 * OR ANY PART THEREOF.
 *
 * In no event will Sun Microsystems, Inc. be liable for any lost revenue
 * or profits or other special, indirect and consequential damages, even if
 * Sun has been advised of the possibility of such damages.
 *
 * Sun Microsystems, Inc.
 * 2550 Garcia Avenue
 * Mountain View, California  94043
 */

/*
 * g711.c
 *
 * u-law, A-law and linear PCM conversions.
 */

/*
 * December 30, 1994:
 * Functions linear2alaw, linear2ulaw have been updated to correctly
 * convert unquantized 16 bit values.
 * Tables for direct u- to A-law and A- to u-law conversions have been
 * corrected.
 * Borge Lindberg, Center for PersonKommunikation, Aalborg University.
 * bli@cpk.auc.dk
 *
 */
 
#define	SIGN_BIT	(0x80)		/* Sign bit for a A-law byte. */
#define	QUANT_MASK	(0xf)		/* Quantization field mask. */
#define	NSEGS		(8)		/* Number of A-law segments. */
#define	SEG_SHIFT	(4)		/* Left shift for segment number. */
#define	SEG_MASK	(0x70)		/* Segment field mask. */

static int seg_aend[8] = {0x1F, 0x3F, 0x7F, 0xFF,
			    0x1FF, 0x3FF, 0x7FF, 0xFFF};
static int seg_uend[8] = {0x3F, 0x7F, 0xFF, 0x1FF,
			    0x3FF, 0x7FF, 0xFFF, 0x1FFF};

#if 0

/* copy from CCITT G.711 specifications */
unsigned char u2a[128] = {			/* u- to A-law conversions */
	1,	1,	2,	2,	3,	3,	4,	4,
	5,	5,	6,	6,	7,	7,	8,	8,
	9,	10,	11,	12,	13,	14,	15,	16,
	17,	18,	19,	20,	21,	22,	23,	24,
	25,	27,	29,	31,	33,	34,	35,	36,
	37,	38,	39,	40,	41,	42,	43,	44,
	46,	48,	49,	50,	51,	52,	53,	54,
	55,	56,	57,	58,	59,	60,	61,	62,
	64,	65,	66,	67,	68,	69,	70,	71,
	72,	73,	74,	75,	76,	77,	78,	79,
/* corrected:
	81,	82,	83,	84,	85,	86,	87,	88, 
   should be: */
	80,	82,	83,	84,	85,	86,	87,	88,
	89,	90,	91,	92,	93,	94,	95,	96,
	97,	98,	99,	100,	101,	102,	103,	104,
	105,	106,	107,	108,	109,	110,	111,	112,
	113,	114,	115,	116,	117,	118,	119,	120,
	121,	122,	123,	124,	125,	126,	127,	128};

unsigned char a2u[128] = {			/* A- to u-law conversions */
	1,	3,	5,	7,	9,	11,	13,	15,
	16,	17,	18,	19,	20,	21,	22,	23,
	24,	25,	26,	27,	28,	29,	30,	31,
	32,	32,	33,	33,	34,	34,	35,	35,
	36,	37,	38,	39,	40,	41,	42,	43,
	44,	45,	46,	47,	48,	48,	49,	49,
	50,	51,	52,	53,	54,	55,	56,	57,
	58,	59,	60,	61,	62,	63,	64,	64,
	65,	66,	67,	68,	69,	70,	71,	72,
/* corrected:
	73,	74,	75,	76,	77,	78,	79,	79,
   should be: */
	73,	74,	75,	76,	77,	78,	79,	80,

	80,	81,	82,	83,	84,	85,	86,	87,
	88,	89,	90,	91,	92,	93,	94,	95,
	96,	97,	98,	99,	100,	101,	102,	103,
	104,	105,	106,	107,	108,	109,	110,	111,
	112,	113,	114,	115,	116,	117,	118,	119,
	120,	121,	122,	123,	124,	125,	126,	127};

#endif

static int
search(
	int		val,	/* changed from "short" *drago* */
	int *	table,
	int		size)	/* changed from "short" *drago* */
{
	int		i;		/* changed from "short" *drago* */

	for (i = 0; i < size; i++) {
		if (val <= *table++)
			return (i);
	}
	return (size);
}

/*
 * linear2alaw() - Convert a 16-bit linear PCM value to 8-bit A-law
 *
 * linear2alaw() accepts an 16-bit integer and encodes it as A-law data.
 *
 *		Linear Input Code	Compressed Code
 *	------------------------	---------------
 *	0000000wxyza			000wxyz
 *	0000001wxyza			001wxyz
 *	000001wxyzab			010wxyz
 *	00001wxyzabc			011wxyz
 *	0001wxyzabcd			100wxyz
 *	001wxyzabcde			101wxyz
 *	01wxyzabcdef			110wxyz
 *	1wxyzabcdefg			111wxyz
 *
 * For further information see John C. Bellamy's Digital Telephony, 1982,
 * John Wiley & Sons, pps 98-111 and 472-476.
 */
int linear2alaw(int	pcm_val)        /* 2's complement (16-bit range) */
                                        /* changed from "short" *drago* */
{
	int		mask;	/* changed from "short" *drago* */
	int		seg;	/* changed from "short" *drago* */
	int		aval;

	pcm_val = pcm_val >> 3;

	if (pcm_val >= 0) {
		mask = 0xD5;		/* sign (7th) bit = 1 */
	} else {
		mask = 0x55;		/* sign bit = 0 */
		pcm_val = -pcm_val - 1;
	}

	/* Convert the scaled magnitude to segment number. */
	seg = search(pcm_val, seg_aend, 8);

	/* Combine the sign, segment, and quantization bits. */

	if (seg >= 8)		/* out of range, return maximum value. */
		return (0x7F ^ mask);
	else {
		aval = seg << SEG_SHIFT;
		if (seg < 2)
			aval |= (pcm_val >> 1) & QUANT_MASK;
		else
			aval |= (pcm_val >> seg) & QUANT_MASK;
		return (aval ^ mask);
	}
}

/*
 * alaw2linear() - Convert an A-law value to 16-bit linear PCM
 *
 */
int alaw2linear(int	a_val)		
{
	int		t;      /* changed from "short" *drago* */
	int		seg;    /* changed from "short" *drago* */

	a_val ^= 0x55;

	t = (a_val & QUANT_MASK) << 4;
	seg = ((unsigned)a_val & SEG_MASK) >> SEG_SHIFT;
	switch (seg) {
	case 0:
		t += 8;
		break;
	case 1:
		t += 0x108;
		break;
	default:
		t += 0x108;
		t <<= seg - 1;
	}
	return ((a_val & SIGN_BIT) ? t : -t);
}

#define	BIAS		(0x84)		/* Bias for linear code. */
#define CLIP            8159

/*
 * linear2ulaw() - Convert a linear PCM value to u-law
 *
 * In order to simplify the encoding process, the original linear magnitude
 * is biased by adding 33 which shifts the encoding range from (0 - 8158) to
 * (33 - 8191). The result can be seen in the following encoding table:
 *
 *	Biased Linear Input Code	Compressed Code
 *	------------------------	---------------
 *	00000001wxyza			000wxyz
 *	0000001wxyzab			001wxyz
 *	000001wxyzabc			010wxyz
 *	00001wxyzabcd			011wxyz
 *	0001wxyzabcde			100wxyz
 *	001wxyzabcdef			101wxyz
 *	01wxyzabcdefg			110wxyz
 *	1wxyzabcdefgh			111wxyz
 *
 * Each biased linear code has a leading 1 which identifies the segment
 * number. The value of the segment number is equal to 7 minus the number
 * of leading 0's. The quantization interval is directly available as the
 * four bits wxyz.  * The trailing bits (a - h) are ignored.
 *
 * Ordinarily the complement of the resulting code word is used for
 * transmission, and so the code word is complemented before it is returned.
 *
 * For further information see John C. Bellamy's Digital Telephony, 1982,
 * John Wiley & Sons, pps 98-111 and 472-476.
 */
int linear2ulaw( int	pcm_val)	/* 2's complement (16-bit range) */
{
	int		mask;
	int		seg;
	int		uval;

	/* Get the sign and the magnitude of the value. */
	pcm_val = pcm_val >> 2;
	if (pcm_val < 0) {
		pcm_val = -pcm_val;
		mask = 0x7F;
	} else {
		mask = 0xFF;
	}
        if ( pcm_val > CLIP ) pcm_val = CLIP;		/* clip the magnitude */
	pcm_val += (BIAS >> 2);

	/* Convert the scaled magnitude to segment number. */
	seg = search(pcm_val, seg_uend, 8);

	/*
	 * Combine the sign, segment, quantization bits;
	 * and complement the code word.
	 */
	if (seg >= 8)		/* out of range, return maximum value. */
		return (0x7F ^ mask);
	else {
		uval = (seg << 4) | ((pcm_val >> (seg + 1)) & 0xF);
		return (uval ^ mask);
	}

}

/*
 * ulaw2linear() - Convert a u-law value to 16-bit linear PCM
 *
 * First, a biased linear code is derived from the code word. An unbiased
 * output can then be obtained by subtracting 33 from the biased code.
 *
 * Note that this function expects to be passed the complement of the
 * original code word. This is in keeping with ISDN conventions.
 */
int ulaw2linear( int	u_val)
{
	int t;

	/* Complement to obtain normal u-law value. */
	u_val = ~u_val;

	/*
	 * Extract and bias the quantization bits. Then
	 * shift up by the segment number and subtract out the bias.
	 */
	t = ((u_val & QUANT_MASK) << 3) + BIAS;
	t <<= (u_val & SEG_MASK) >> SEG_SHIFT;

	return ((u_val & SIGN_BIT) ? (BIAS - t) : (t - BIAS));
}

#if 0

/* A-law to u-law conversion */
static int alaw2ulaw (int	aval)
{
	aval &= 0xff;
	return ((aval & 0x80) ? (0xFF ^ a2u[aval ^ 0xD5]) :
	    (0x7F ^ a2u[aval ^ 0x55]));
}

/* u-law to A-law conversion */
static int ulaw2alaw (int	uval)
{
	uval &= 0xff;
	return ((uval & 0x80) ? (0xD5 ^ (u2a[0xFF ^ uval] - 1)) :
	    (0x55 ^ (u2a[0x7F ^ uval] - 1)));
}

#endif

